src/profile-logic/symbolication.js

```diff
@@ -640,64 +726,146 @@ export function applySymbolicationStep(
-        funcTable.length++;
+        funcIndex = availableFuncIter.next().value;
+        if (funcIndex === undefined) {
+          // Need a new func.
+          funcIndex = funcTable.length;
+          funcTable.isJS[funcIndex] = false;
+          funcTable.relevantForJS[funcIndex] = false;
+          funcTable.resource[funcIndex] = resourceIndex;
+          funcTable.fileName[funcIndex] = null;
+          funcTable.lineNumber[funcIndex] = null;
+          funcTable.columnNumber[funcIndex] = null;
+          // The name field will be filled below.
+          funcTable.length++;
+        }
+        funcTable.name[funcIndex] = functionStringIndex;
+        funcTable.fileName[funcIndex] = fileNameStringIndex;
+        funcKeyToFuncMap.set(funcKey, funcIndex);
       }
-      funcTable.name[funcIndex] = functionStringIndex;
-      funcTable.fileName[funcIndex] = fileNameStringIndex;
-      funcKeyToFuncMap.set(funcKey, funcIndex);
-    }
+      inlineExpansionFuncIndexes.push(funcIndex);
+      let expansionFrameIndex;
+      if (inlineDepth === 0) {
+        // This is an outer frame.
+        expansionFrameIndex = frameIndex;
+      } else {
+        // This is an inline frame.
+        // Add a frame at this depth. Try to use an unused existing frame, or
+        // create a completely new frame if no frames are available.
+        expansionFrameIndex = availableFrameIter.next().value;
+        if (expansionFrameIndex === undefined) {
+          expansionFrameIndex = frameTable.length;
+          frameTable.length++;
+        }
 
-    frameTable.func[frameIndex] = funcIndex;
-    frameTable.line[frameIndex] = addressResult.line ?? null;
-    oldFuncToNewFuncEntries.push([oldFunc, funcIndex]);
+        // Copy most fields over from the outer frame, unchanged.
+        const category = frameTable.category[frameIndex];
+        const subcategory = frameTable.subcategory[frameIndex];
+        const innerWindowID = frameTable.innerWindowID[frameIndex];
+        const implementation = frameTable.implementation[frameIndex];
+        const optimizations = frameTable.optimizations[frameIndex];
+        frameTable.category[expansionFrameIndex] = category;
+        frameTable.subcategory[expansionFrameIndex] = subcategory;
+        frameTable.innerWindowID[expansionFrameIndex] = innerWindowID;
+        frameTable.implementation[expansionFrameIndex] = implementation;
+        frameTable.optimizations[expansionFrameIndex] = optimizations;
+        frameTable.address[expansionFrameIndex] = address;
+        frameTable.nativeSymbol[expansionFrameIndex] = nativeSymbolIndex;
+
+        // These remaining fields are filled below.
+      }
+      frameTable.inlineDepth[expansionFrameIndex] = inlineDepth;
+      frameTable.func[expansionFrameIndex] = funcIndex;
+      frameTable.line[expansionFrameIndex] = frameInfo.line ?? null;
+      frameTable.column[expansionFrameIndex] = null;
+      inlineExpansionFrames.push(expansionFrameIndex);
```

Comment:
I wonder if this should be a function.

Reply:
Hmm, I don't see a clean way to break this out into its own function. I can take another look after this is merged.